src/internals/objects/impl/objects.ts

```diff
@@ -79,9 +86,31 @@ type RecursiveGet<Obj, pathList> = Obj extends any
     : Obj
   : never;
 
-export type PartialDeep<T> = T extends object
-  ? { [P in keyof T]?: PartialDeep<T[P]> }
-  : T;
+export type PartialDeep<Type> = Type extends Function | Date
+  ? Type
+  : Type extends Map<infer Keys, infer Values>
+  ? Map<PartialDeep<Keys>, PartialDeep<Values>>
+  : Type extends ReadonlyMap<infer Keys, infer Values>
+  ? ReadonlyMap<PartialDeep<Keys>, PartialDeep<Values>>
+  : Type extends WeakMap<infer Keys, infer Values>
+  ? WeakMap<PartialDeep<Keys>, PartialDeep<Values>>
+  : Type extends Set<infer Values>
+  ? Set<PartialDeep<Values>>
+  : Type extends ReadonlySet<infer Values>
+  ? ReadonlySet<PartialDeep<Values>>
+  : Type extends WeakSet<infer Values>
+  ? WeakSet<PartialDeep<Values>>
+  : Type extends Array<infer Values>
+  ? IsTuple<Type> extends true
+    ? { [Key in keyof Type]?: PartialDeep<Type[Key]> }
+    : Array<PartialDeep<Values> | undefined>
+  : Type extends Promise<infer Value>
+  ? Promise<PartialDeep<Value>>
+  : Type extends object
+  ? { [Key in keyof Type]?: PartialDeep<Type[Key]> }
+  : IsUnknown<Type> extends true
+  ? unknown
+  : Partial<Type>;
```

Comment:
Source of changes – https://github.com/ts-essentials/ts-essentials/blob/master/lib/types.ts#L36-L60